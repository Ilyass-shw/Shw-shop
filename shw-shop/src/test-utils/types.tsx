import { configureStore, createSlice } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';

export type StoreType = ReturnType<typeof configureStore>;
export type RenderReturn = ReturnType<typeof render>;
export type SliceParameters = Parameters<typeof createSlice>;
export type UiType = React.ReactElement;

export interface WrapperParams {
  MyStore: StoreType;
  WithRouter?: boolean;
}

export interface useFormProps {
  email: string;
  phone: number;
  buyerAcceptMarketing: string;
}
