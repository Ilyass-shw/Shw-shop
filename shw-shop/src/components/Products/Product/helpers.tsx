import { isInViewport } from '../../../helpers/isInViewport';

export const handleScroll = (
  img: Element | null | undefined,
  setIsVisible: React.Dispatch<React.SetStateAction<boolean>>,
): void => {
  // we need to make sure img.current in defined
  // before using it in isInViewport as the latter
  // accepts type Element only
  if (img) {
    setIsVisible(isInViewport(img));
  }
};
