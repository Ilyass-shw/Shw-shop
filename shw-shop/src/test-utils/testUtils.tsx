import React, { ReactElement } from 'react';
import { render as rtlRender, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import store, { makeCustomTestStore } from './faksestore';
import { RenderReturn, StoreType, useFormProps } from './types';
import { BrowserRouter } from 'react-router-dom';
import { realStoreType } from '../App/store';
import { useForm, FormProvider } from 'react-hook-form';

export const render = (ui: React.ReactElement): RenderReturn => {
  return rtlRender(ui);
};

// ============= ============= ============= ============= =============

export const renderWithStore = (
  ui: React.ReactElement,
  MyStore: realStoreType,
  WithRouter?: 'withRouter',
): RenderReturn => {
  if (WithRouter) {
    return rtlRender(
      <Provider store={MyStore}>
        <BrowserRouter>{ui}</BrowserRouter>
      </Provider>,
    );
  } else {
    return rtlRender(<Provider store={MyStore}>{ui}</Provider>);
  }
};

// ============= ============= ============= ============= =============

export const renderWithRouter = (ui: React.ReactElement): RenderReturn => {
  return rtlRender(<BrowserRouter>{ui}</BrowserRouter>);
};

// ============= ============= ============= ============= =============

export const makeTestStore = (
  Name?: string,
  initialState?: Record<string, unknown>,
  reducers = {},
): StoreType => {
  if (Name && initialState) {
    const CustomTestStore = makeCustomTestStore(Name, initialState, reducers);
    const origDispatch = CustomTestStore.dispatch;
    CustomTestStore.dispatch = jest.fn(origDispatch);
    return CustomTestStore;
  } else {
    const origDispatch = store.dispatch as jest.Mock;
    store.dispatch = jest.fn(origDispatch);
    return store;
  }
};

// re-export everything
export * from '@testing-library/react';

// ============= ============= ============= ============= =============

export const renderWithFormProvider = (
  ui: ReactElement,
  WithRouter?: 'withRouter',
  { defaultValues = {} } = {},
): RenderResult => {
  // let reactHookFormMethods = {};

  const Wrapper: React.FC = ({ children }) => {
    const methods = useForm({ defaultValues });
    return <FormProvider {...methods}>{children}</FormProvider>;
  };
  const WithRouterWrapper: React.FC = ({ children }) => {
    const methods = useForm({ defaultValues });
    return (
      <BrowserRouter>
        <FormProvider {...methods}>{children}</FormProvider>
      </BrowserRouter>
    );
  };

  if (!WithRouter) {
    return rtlRender(ui, { wrapper: Wrapper });
  }
  return rtlRender(ui, { wrapper: WithRouterWrapper });
};

// ============= ============= ============= ============= =============

export const renderWithReactHookForm = (
  ui: JSX.Element,
  saveData: jest.Mock,
): RenderResult => {
  const Wrapper: React.FC = ({ children }) => {
    const methods = useForm<useFormProps>();
    return (
      <FormProvider {...methods}>
        <form
          onSubmit={methods.handleSubmit((data) => {
            saveData(data);
          })}
        >
          <BrowserRouter>{children}</BrowserRouter>
          <button type="submit" data-testid="testSubmitButton"></button>
        </form>
      </FormProvider>
    );
  };
  return rtlRender(<Wrapper>{ui}</Wrapper>);
};
