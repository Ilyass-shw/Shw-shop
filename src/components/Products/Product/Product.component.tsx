import styled from 'styled-components';
import { RouterLink } from '../../../App/App.component';

export const ProductItem = styled.li`
  overflow: hidden;
  padding: 1rem 1rem;
  width: 85vw;
  margin: auto;
  line-height: 1.8rem;
  @media screen and (min-width: 960px) {
    width: 40vw;
  }
  @media screen and (min-width: 1276px) {
    width: 30vw;
  }
`;

// ============= ============= ============= ============= =============

export const ProductLink = styled(RouterLink)`
  display: flex;
  flex-direction: column;
  align-items: center;
  color: black;
`;

// ============= ============= ============= ============= =============

export const ItemName = styled.h4`
  padding: 0.2rem 0 0 0;
  text-align: center;
`;

// ============= ============= ============= ============= =============

export const ItemPrice = styled.h6`
  font-weight: 600;
`;
