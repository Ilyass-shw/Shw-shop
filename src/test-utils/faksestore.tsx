import { configureStore } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import { StoreType } from './types';

const fakeSlice = createSlice({
  name: 'name',
  initialState: {},
  reducers: {},
});

export default configureStore({
  reducer: {
    posts: fakeSlice.reducer,
  },
});

export const makeCustomTestStore = (
  Name: string,
  initialState: Record<string, unknown>,
  reducers = {},
): StoreType => {
  const FakeSlice = createSlice({
    name: Name,
    initialState,
    reducers,
  });

  return configureStore({
    reducer: {
      posts: FakeSlice.reducer,
    },
  });
};
