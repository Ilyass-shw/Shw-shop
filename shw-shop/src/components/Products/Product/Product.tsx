import React, { useRef, useState, useEffect } from 'react';
import ProductImgs from '../ProductImgs/ProductImgs';
import { Item } from '../../../helpers/types';
import { handleScroll } from './helpers';
import {
  ProductItem,
  ItemName,
  ItemPrice,
  ProductLink,
} from './Product.component';

interface product {
  item: Item;
}
const Product: React.FC<product> = ({ item }) => {
  const [isHover, setIsHover] = useState(false);
  const [isVisible, setIsVisible] = useState(false);
  const img = useRef<HTMLAnchorElement>(null);

  useEffect(() => {
    window.addEventListener('scroll', () =>
      handleScroll(img.current, setIsVisible),
    );
    return () => {
      window.removeEventListener('scroll', () =>
        handleScroll(img.current, setIsVisible),
      );
    };
  }, []);
  return (
    <ProductItem
      onMouseOver={(): void => setIsHover(true)}
      onMouseOut={(): void => setIsHover(false)}
    >
      <ProductLink
        ref={img}
        to={`/product/${item.id}`}
        aria-label="product card"
      >
        <ProductImgs
          imgs={item.imgs}
          alt={item.name}
          isOnHover={isHover}
          isVisible={isVisible}
        />
        <ItemName>{item.name}</ItemName>
        <ItemPrice>{'$' + item.price}</ItemPrice>
      </ProductLink>
    </ProductItem>
  );
};

export default Product;
