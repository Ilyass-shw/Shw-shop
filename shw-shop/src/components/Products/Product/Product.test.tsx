import React from 'react';
import { renderWithRouter } from '../../../test-utils/testUtils';
import Product from './Product';
// import ProductImgs from '../ProductImgs/ProductImgs';

describe('Product', () => {

  const Item = {
    name: 'itemOne',
    id: 1,
    description: 'rand',
    imgs: ['src1', 'src2', 'src3'],
    price: 45,
  };

  it('should render', () => {
    renderWithRouter(<Product item={Item} />);
  });
  it('should match snaptshot', () => {
    const { asFragment } = renderWithRouter(<Product item={Item} />);
    expect(asFragment()).toMatchSnapshot();
  });
//   it('should render ProductImgs with the right isOnHover prop ', () => {
//     // ProductImgs = jest.fn()
//     renderWithRouter(<Product item={Item} />);
//     // const productImgs = screen.getByTestId('productImgs');
//     expect(ProductImgs).toHaveBeenCalledWith('imgs', Item.imgs);
//     expect(ProductImgs).toHaveBeenCalledWith('isOnHover', false);
//     expect(ProductImgs).toHaveBeenCalledWith('isVisible', false);
//   });
});
