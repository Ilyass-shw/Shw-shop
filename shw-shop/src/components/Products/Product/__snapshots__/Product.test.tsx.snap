// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should match snaptshot 1`] = `
<DocumentFragment>
  <li
    class="sc-gKckTs jFUhbv"
  >
    <a
      aria-label="product card"
      class="sc-eCImvq sc-iCfLBT gECYPg bNwGWt"
      href="/product/1"
    >
      <div>
        <div
          class="sc-bdvvaa cYloRB"
          data-testid="product-images"
        >
          <img
            alt="itemOne"
            class="sc-gsDJrp dlZrhb"
            loading="lazy"
            src="src1"
          />
          <img
            alt="itemOne"
            class="sc-gsDJrp dmjAPj"
            loading="lazy"
            src="src2"
          />
          <img
            alt="itemOne"
            class="sc-gsDJrp fGGvVI"
            loading="lazy"
            src="src3"
          />
        </div>
        <div
          class="sc-dkPtyc goXLUj"
        >
          <div
            class="sc-hKwCoD kLaVan"
          />
        </div>
      </div>
      <h4
        class="sc-furvIG boMFUg"
      >
        itemOne
      </h4>
      <h6
        class="sc-pVTma bqGpcZ"
      >
        $45
      </h6>
    </a>
  </li>
</DocumentFragment>
`;
